File: leetcode_208/java/Trie.java
/**
 * Trie（发音叫"try"）或者叫前缀树是一种树数据结构，用于检索字符串数据集中的键
 */
class TrieNode {

  private TrieNode[] links;

  private final int R = 26;

  // 表示节点是键的结尾还是键前缀
  private boolean isEnd;

  public TrieNode() {
    links = new TrieNode[R];
  }

  public boolean containsKey(char ch) {
    return links[ch - 'a'] != null;
  }

  public TrieNode get(char ch) {
    return links[ch - 'a'];
  }

  public void put(char ch, TrieNode node) {
    links[ch - 'a'] = node;
  }

  public void setEnd() {
    isEnd = true;
  }

  public boolean isEnd() {
    return isEnd;
  }
}

class Trie {
  private TrieNode root;

  public Trie() {
    root = new TrieNode();
  }

  public void insert(String word) {
    TrieNode node = root;

    for (int i = 0; i < word.length(); i++) {
      char currentChar = word.charAt(i);
      if (!node.containsKey(currentChar)) {
        node.put(currentChar, new TrieNode());
      }
      node = node.get(currentChar);
    }

    node.setEnd();
  }

  private TrieNode searchPrefix(String word) {
    TrieNode node = root;

    for (int i = 0; i < word.length(); i++) {
      char curLetter = word.charAt(i);

      if (node.containsKey(curLetter)) {
        node = node.get(curLetter);
      } else {
        return null;
      }
    }

    return node;
  }

  public boolean search(String word) {
    TrieNode node = searchPrefix(word);

    return node != null && node.isEnd();
  }

  public boolean startsWith(String prefix) {
    TrieNode node = searchPrefix(prefix);

    return node != null;
  }
}

File: leetcode_208/Trie.js
/**
* 定义前缀树节点
*/
class TrieNode {
  constructor() {
    this.links = new Map();
    this.isEnd = false;
  }

  containsKey(ch) {
    return this.links.has(ch);
  }

  get(ch) {
    return this.links.get(ch);
  }

  put(ch, node) {
    this.links.set(ch, node);
  }

  setEnd() {
    this.isEnd = true;
  }

  isEndNode() {
    return this.isEnd;
  }
}

/**
  * 实现前缀树
  */
class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  /**
   * 在前缀树中搜索指定的前缀
   * @param {string} word 待搜索的键
   * @return {null || TrieNode}
   */
  searchPrefix(word) {
    let currentNode = this.root;

    for (const ch of word) {
      // 如果当前节点的Map中不存在当前要查找的字符，说明键在前缀树中不存在，搜索结束
      if (!currentNode.links.has(ch)) {
        return null;
      }

      currentNode = currentNode.get(ch);
    }

    return currentNode;
  }

  /**
   * 查找指定的键
   * @param {string} word 指定的键
   * @return {boolean}
   */
  search(word) {
    const node = this.searchPrefix(word);

    // 当searchPrefix函数的返回值是代表结束的TrieNode时，表示查找到了完整的键
    return (node !== null) && node.isEndNode();
  }

  /**
   * 在前缀树中查找键前缀
   * @param {string} prefix 指定的前缀
   * @return {boolean}
   */
  startsWith(prefix) {
    const node = this.searchPrefix(prefix);

    return node !== null;
  }

  /**
  * 插入指定的键
  * @param {string} word 待插入的键
  */
  insert(word) {
    let currentNode = this.root;
    for (const ch of word) {
      // 如果当前节点不包含当前迭代的字符，就以该字符为key，创建一个新的节点为value，
      // 将该键值对插入Map
      if (!currentNode.links.has(ch)) {
        currentNode.put(ch, new TrieNode());
      }
      
      // 移动指针，指向字符对应的节点
      currentNode = currentNode.get(ch);
    }
    
    // 设置最后一个节点的end标志，表示插入结束
    currentNode.setEnd();
  }
}
